### ISSUE TITLE
Improved support for command-line arguments (eg in help)

### ISSUE THREAD

**Issue by johncrim** [adamralph/bullseye#354]
## Use case(s)

Ideal use case: An app with multiple targets, where each target can have target-specific parameters. Some parameters may be global or shared between targets.

I'd settle for: Global parameters only (like https://github.com/adamralph/bullseye#command-line-arguments examples), but argument help text is included in output of `build.sh --help` and/or `build.sh --list-targets`

## Description

The info provided here: https://github.com/adamralph/bullseye#command-line-arguments provides a start, but it seems a bit frankenstein-y trying to make command-line processing libraries work with BullsEye. We have 2 different libraries doing their own command-line parsing, and building up their own respective internal models from the arguments.

One problem that arises: How do we make help work, such that available parameters are listed?

Another problem: I want to structure multiple targets, with dependencies, but different targets have different required parameters. When a target is run, each of its dependencies are run, and the parameters for those dependencies are required.

Libraries like System.CommandLine.Experimental, and McMaster.Extensions.CommandLineUtils can support command-specific parameters, but they don't have the dependency support, or the nice build-style output that Bullseye has.

I realize this is a "big" feature request. It might require doing the target management with BullsEye, and the command-line processing using another library.

## Alternatives

In the absence of first-class support for this, a better example that includes (target-specific?) parameters in the help output would be useful. I'm happy to help with this, but right now I'm kind of stuck.

**Comment by adamralph** [comment]
@johncrim thanks for raising this.

My current thinking is this: Bullseye supplies CLI parsing for the basic scenario. That is, when you don't need to define any custom CLI arguments or options, and you just want the out-of-the-box behaviour from Bullseye. For a more advanced scenario, you need to take control of the CLI parsing yourself, including the production of help text. Both `System.CommandLine.Experimental` and `McMaster.Extensions.CommandLineUtils` have built-in help text generation. This advanced scenario also involves translating between your CLI and Bullseye.

Ultimately, you are writing a console app, and Bullseye is just a library that you are calling. Bullseye just happens to provide a facility for you to pass through the CLI args for the basic scenario.

There are two awkward things about the the advanced scenario right now. Firstly, you would have to repeat the Bullseye arguments (list of targets) and option definitions in your app. Secondly, you would have to translate your arguments and options back to an array of strings to pass them to Bullseye. But if Bullseye exposed its option definitions, and accepted named options in `RunTargetsAndExit` overloads, perhaps the advanced scenario I describe would be palatable enough?

The API usage could look something like this (when using `System.CommandLine.Experimental`):

```c#
static int Main(string[] args)
{
    var cmd = new RootCommand();

    cmd.Add(new Option(new[] { "--foo", "-f" }, "A foo.") { Argument = new Argument<string>("foo") });

    // translate from Bullseye to System.CommandLine.Experimental
    cmd.Add(new Argument("targets") { Arity = ArgumentArity.ZeroOrMore });
    foreach (var option in Bullseye.Options.Definitions)
    {
        cmd.Add(new Option(option.Aliases, option.Description) { Argument = new Argument() { ArgumentType = option.OptionType } });
    }

    cmd.Handler = CommandHandler.Create<string>(foo =>
    {
        Target("build", () => System.Console.WriteLine($"foo = {foo}"));

        Target("default", DependsOn("build"));

        // translate from System.CommandLine.Experimental to Bullseye
        var cmdLine = cmd.Parse(args);
        var targets = cmdLine.CommandResult.Tokens.Select(token => token.Value);
        var options = Bullseye.Options.Definitions.Select(option => option.Aliases[0]).Select(alias => (alias, cmdLine.ValueForOption(alias)));

        RunTargetsAndExit(targets, options);
    });

    return cmd.Invoke(args);
}
```

The code looks quite verbose, but most of it is the code that you would need in _any_ console app. There are two pieces of boilerplate required to translate between Bullseye and `System.CommandLine.Experimental`.

```c#
// translate from Bullseye to System.CommandLine.Experimental
cmd.Add(new Argument("targets") { Arity = ArgumentArity.ZeroOrMore });
foreach (var option in Bullseye.Options.Definitions)
{
    cmd.Add(new Option(option.Aliases, option.Description) { Argument = new Argument() { ArgumentType = option.OptionType } });
}
```

and

```c#
// translate from System.CommandLine.Experimental to Bullseye
var cmdLine = cmd.Parse(args);
var targets = cmdLine.CommandResult.Tokens.Select(token => token.Value);
var options = Bullseye.Options.Definitions.Select(option => option.Aliases[0]).Select(alias => (alias, cmdLine.ValueForOption(alias)));
```

This program would already respond to built-in options in `System.CommandLine.Experimental` such as `--help` and `--version`. E.g. `--help` would give something like:

```
Usage:
  targets [options] [<targets>...]

Arguments:
  <targets>

Options:
  -f, --foo <foo>    A foo.
  -c, --clear        Clear the console before execution
  -n, --dry-run      Do a dry run without executing actions
  {...the rest of the Bullseye options...}
  --version          Display version information
```

The API's which are missing right now are `Bullseye.Options.Definitions` and `RunTargetsAndExit(IEnumerable<string> targets, IEnumerable<(string, object)> options)`, so they would have to be added.

Thoughts?

**Comment by johncrim** [comment]
Thank you @adamralph , this looks exactly like what I'm looking for. I agree with your thinking on Bullseye exposing a model for advanced integration where it isn't responsible for command-line processing.

The proposed integration code isn't at all too verbose, IMO - using System.CommandLine.Experimental is somewhat verbose for any non-trivial command-line app. I've been digging into that project, too.

The biggest question your proposal raises for me, is: Is there a way to support target-specific parameters, similar to how System.CommandLine.Experimental can have subcommands, each with their own distinct parameters? This would be like putting `RunTargetsAndExit()` and all the Bullseye options in the `RootCommand`, and defining separate `Command`s associated with `Target()`s.

This might be doable by defining the `Target()`s at the top level, and invoking the commands within them.

Other than that, the only omissions I see in this example are:
* Some of the Bullseye options don't have Arguments, so would need to handle that
* How are validation errors passed back, eg if a non-existent target is specified?

**Comment by adamralph** [comment]
> Is there a way to support target-specific parameters, similar to how System.CommandLine.Experimental can have subcommands, each with their own distinct parameters?

@johncrim do you think it would be enough for you handle that internally in your program? That is, you can choose what to do with the options once your program has them, and which targets use them.

> Some of the Bullseye options don't have Arguments, so would need to handle that

In fact, none of the Bullseye options have arguments. But effectively, that makes them all Boolean, with their argument being `true` or `false`. That way, the example API I showed above will still work, since each option definition will tell you that the option is a Boolean, and when you pass back the named arguments, you will be passing back Boolean values.

> How are validation errors passed back, eg if a non-existent target is specified?

If you call `RunTargetsAndExit`, which is usually what you will want to do, then Bullseye will write error details to `stderr` and terminate the process.

If you call `RunTargetsWithoutExiting`, which only exists for edge cases, then Bullseye will throw an exception which you can handle appropriately.

**Comment by johncrim** [comment]
> In fact, none of the Bullseye options have arguments. But effectively, that makes them all Boolean, with their arguments being true or false. That way, the example API I showed above will still work, since each option definition will tell you that the option is a Boolean, and when you pass back the named arguments, you will be passing back Boolean values.

Makes sense, I wasn't aware of that way to work with boolean Options.

> do you think it would be enough for you handle that internally in your program? That is, you can choose what to do with the options once your program has them, and which targets use them.

To hopefully clarify my thinking, here's a revision of your example. It obviously doesn't compile, but would be great IMO if something like this could be written:

```c#
static int Main(string[] args)
{

    var packageCmd = new Command("package")
          {
             // Output option is only parsed for the package target, or any target that depends on package
             new Option("--output") { Argument = new Argument<DirectoryInfo>("outputDirectory") });
          }
    packageCmd.Handler = CommandHandler.Create<DirectoryInfo>( (outputDirectory) => .. some other logic );

    var cmd = new RootCommand() { packageCmd }
   // Foo option is global (all targets)
    cmd.Add(new Option(new[] { "--foo", "-f" }, "A foo.") { Argument = new Argument<string>("foo") });

    // translate from Bullseye to System.CommandLine.Experimental
    cmd.Add(new Argument("targets") { Arity = ArgumentArity.ZeroOrMore });
    foreach (var option in Bullseye.Options.Definitions)
    {
        cmd.Add(new Option(option.Aliases, option.Description) { Argument = new Argument() { ArgumentType = option.OptionType } });
    }

    // CommandLine parsing has to be done after CommandLine definition
    var cmdLine = cmd.Parse(args); 

    // This probably doesn't work, but something like this might
    Target("package", () => packageCmd.Invoke(cmdLine));

    Target("build", () => System.Console.WriteLine($"foo = {foo}"));

    Target("default", DependsOn("build", "package"));

    cmd.Handler = CommandHandler.Create<string>(foo =>
    {
        // translate from System.CommandLine.Experimental to Bullseye
        var targets = cmdLine.CommandResult.Tokens.Select(token => token.Value);
        var options = Bullseye.Options.Definitions.Select(option => option.Aliases[0]).Select(alias => (alias, cmdLine.ValueForOption(alias)));

        RunTargetsAndExit(targets, options);
    });

    return cmd.Invoke(args);
}
```

What do you think?

**Comment by adamralph** [comment]
@johncrim I think it's best to carve this out into three seperate enhancements:

- Expose option definitions in API (so that you can translate them to {your command line lib of choice})
- Provide RunTargets... overloads accepting a list of targets and a dictionary of options (to make it easier to translate from {your command line lib of choice})
- Something around subcommands and targets

The first two are an easy sell. I'm happy to add them in 3.1.

The third, I'm not yet sure about. My current thinking is that Bullseye doesn't really have part to play in it. Using {your command line lib of choice} you can form whatever subcommands, arguments, and options you like. So long as you map them back to Bullseye's list of targets and dictionary of options, everything will just work. Bear in mind that the body of a Bullseye target is arbitrary. That is, this will work today:

```c#
    // This probably doesn't work, but something like this might
    Target("package", () => packageCmd.Invoke(cmdLine));
```

Bullseye doesn't care what `packageCmd` and `cmdLine` are. It will just invoke the delegate, and the runtime will capture locals, etc.

**Comment by johncrim** [comment]
@adamralph - I agree with your structure, and the first 2 would be a nice addition.

I don't know that anything additional is required for subcommands. I just wanted to demonstrate a use-case that supports arguments for individual targets, for your consideration.

I would be happy to test out the first 2 enhancements and provide feedback.

**Comment by adamralph** [comment]
@johncrim FYI I'm working on the first two enhancements in https://github.com/adamralph/bullseye/pull/361. The API I proposed above is in place, but I'm still investigating how this might fit in with [System.CommandLine.DragonFruit](https://www.nuget.org/packages/System.CommandLine.DragonFruit). For simplicity, I decided to collapse the two enhancements into one: "API for integration with third party CLI libs"

Given that it's likely that nothing further is needed for subcommands, I'm closing this issue, and we can continue any further conversation in https://github.com/adamralph/bullseye/pull/361.

Thanks again for raising this.
